### PR TITLE
Use watch instead of list pods in node controller

### PR DESCRIPTION
**Pull request by Sindica**
Reduced # of list pods requests to ETCD. 1 request per 3 second, 1 call per node. This change significantly reduced system load during large cluster density test.

### CODE REVIEW

**Comment by Sindica** [comment]
Can someone review this PR?

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 1]
```diff
@@ -373,21 +374,46 @@ func NewNodeLifecycleController(
...
+					}
+					if len(pod.Spec.NodeName) == 0 {
+						return []string{}, nil
+					}
+					return []string{pod.Spec.NodeName}, nil
+				},
+			})
+
+			podIndexer := podInformer.Informer().GetIndexer()
+			//podLister := podInformer.Lister()
```
Comment (yb01) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go:393]: i think we can just remove those commented-out-code

**Comment by yb01** [comment]
can we test this change with a 1tp 2 rp cluster with smaller number of nodes?

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 2]
```diff
@@ -431,6 +457,15 @@ func NewNodeLifecycleController(
 	nc.nodeLister = nodeInformer.Lister()
 	nc.nodeInformerSynced = nodeInformer.Informer().HasSynced
 
+	nc.podInformersSynced = func() bool {
```
Comment (yb01) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go:460]: the original k8s PR #79370 doesn't have this code to enforce the podInformers synced. any reason we added here?

**Comment by yb01** [comment]
/lgtm.
please run a small cluster test with 1tp 2rp to validate the functionality of the change in scaleout cases.

**Review by yb01 — APPROVED** [review]
/lgtm

**Comment by Sindica** [comment]
> /lgtm.
> please run a small cluster test with 1tp 2rp to validate the functionality of the change in scaleout cases.

I tested in 2tp2rp 2x100 environment, perf test succeeded. However, we got unexpected error in KCM:
E0802 20:59:05.761554       1 taint_manager.go:466] Index with name spec.nodeName does not exist

This happened in 1tp1rp test as well. I will investigate. Otherwise, perf test works fine.

**Comment by Sindica** [comment]
> > /lgtm.
> > please run a small cluster test with 1tp 2rp to validate the functionality of the change in scaleout cases.
> 
> I tested in 2tp2rp 2x100 environment, perf test succeeded. However, we got unexpected error in KCM:
> E0802 20:59:05.761554 1 taint_manager.go:466] Index with name spec.nodeName does not exist
> 
> This happened in 1tp1rp test as well. I will investigate. Otherwise, perf test works fine.

The error seems happen only when cluster start up, not 100% occur, seems depends on the initialization time of the indexer and referral. And only once per node. Will add a log into POC branch to make sure it does not affect the actual index function.

**Comment by Sindica** [comment]
> > > /lgtm.
> > > please run a small cluster test with 1tp 2rp to validate the functionality of the change in scaleout cases.
> > 
> > 
> > I tested in 2tp2rp 2x100 environment, perf test succeeded. However, we got unexpected error in KCM:
> > E0802 20:59:05.761554 1 taint_manager.go:466] Index with name spec.nodeName does not exist
> > This happened in 1tp1rp test as well. I will investigate. Otherwise, perf test works fine.
> 
> The error seems happen only when cluster start up, not 100% occur, seems depends on the initialization time of the indexer and referral. And only once per node. Will add a log into POC branch to make sure it does not affect the actual index function.

Index was not added onto podinformer from tenant managers because informer was started before adding index. Rearranged informer start time, density test on 1x1x100 cluster. Index was properly added and expected debug log appeared.

**Review by yb01 — APPROVED** [review]
/lgtm

**Comment by yb01** [comment]
re-approve, the most recent scale test runs fine

**Pull request merged**